src/VisualStudio/Core/Def/Implementation/Diagnostics/VisualStudioDiagnosticAnalyzerProvider.WorkspaceEventListener.cs

```diff
@@ -62,8 +63,15 @@ private async Task InitializeWorkspaceAsync(VisualStudioWorkspaceImpl workspace)
                     var references = provider.GetAnalyzerReferencesInExtensions();
                     LogWorkspaceAnalyzerCount(references.Length);
 
-                    workspace.ApplyChangeToWorkspace(w =>
-                        w.SetCurrentSolution(s => s.WithAnalyzerReferences(references), WorkspaceChangeKind.SolutionChanged));
+                    if (workspace is VisualStudioWorkspaceImpl vsWorkspace)
+                    {
+                        vsWorkspace.ApplyChangeToWorkspace(w =>
+                            w.SetCurrentSolution(s => s.WithAnalyzerReferences(references), WorkspaceChangeKind.SolutionChanged));
+                    }
+                    else
+                    {
+                        workspace.SetCurrentSolution(s => s.WithAnalyzerReferences(references), WorkspaceChangeKind.SolutionChanged);
+                    }
```

Comment:
What's wrong with pattern matching? Seems like overkill to me to avoid one condition.

Reply:
Basically this is coupling an implementation and threading impact in a way that really shouldn't be. If C# supported some concept of a friend/internal method where ApplyChangeToWorkspace could only be called from certain other names (namely VisualStudioProject) that would have been used here.

Reply:
The call to ApplyChangeToWorkspace was already there. I agree it would be better if it wasn't necessary. The InteractiveWorkspace doesn't need it.

Reply:
Should I file an issue to follow up?

Reply:
Why not just fix it?

Reply:
I'm not sure what you mean by "fix". Do you mean the refactoring to a service as you suggested [above](https://github.com/dotnet/roslyn/pull/49601#discussion_r553682302) or that ApplyChangeToWorkspace should not be called at all?

Reply:
More specifically, was the code wrong/not ideal before this PR?

Reply:
So ApplyChangeToWorkspace is fine to call _somewhere_ but it shouldn't be this component having to magically know the threading implementation details of other components. If this is put behind a workspace service where the workspace can do that, at least it's the workspace owning it's own threading behavior rather than this other component having to know.

It might also be possible to get rid of the ApplyChangeToWorkspace in favor of your SetCurrentSolution model, but right now that we have two competing threading synchronization primitives _and other code has to know this and randomly do different things_ is really icky.

Reply:
Ok, I understand now. I'm not pushing against fixing it, I definitely don't like the current situation. But I'd prefer doing so in a different PR since it's unrelated to interactive window. The problem of fixing it by defining another way of updating solution of a workspace (either a workspace service that applies either a generic transformation to a solution or just a specific one for updating analyzer references) is that one needs to know to use it instead of using `SetCurrentSolution` and `ApplyChangeToWorkspace`. Imo, the right fix is to eliminate `ApplyChangeToWorkspace` and that is certainly out of scope of the PR.

Reply:
Implemented

Reply:
Thanks!